File: packages/gh-tasks/src/commands/plan.ts
import { resolveLocale, t } from '../i18n/index.ts';
import type { AppConfig } from '../lib/config.ts';
import {
  createClient,
  createRestClient,
  type GraphQLClient,
  type RestClient,
  resolveToken,
} from '../lib/github.ts';
import { type Period, parsePeriodFlag, rangeOf, suggestMilestoneTitle } from '../lib/period.ts';
import { ProjectError, type ProjectRef, resolveProjectRef } from '../lib/project.ts';
import { resolveProjectNodeId } from '../lib/projectItem.ts';
import {
  createMilestone,
  LIST_MILESTONES,
  LIST_PROJECT_V2_FIELDS,
  LIST_PROJECT_V2_ITEMS,
  LIST_REPO_ISSUES_WITH_MILESTONE,
  type ListMilestonesResponse,
  type ListProjectV2FieldsResponse,
  type ListProjectV2ItemsResponse,
  type ListRepoIssuesWithMilestoneResponse,
  type ProjectV2FieldNode,
  type ProjectV2ItemNode,
  type ProjectV2IterationOption,
  type RepoIssueWithMilestoneNode,
  UPDATE_ISSUE_MILESTONE,
  UPDATE_PROJECT_V2_ITEM_FIELD_VALUE,
  type UpdateIssueMilestoneResponse,
  type UpdateProjectV2ItemFieldValueResponse,
} from '../lib/queries/index.ts';
import { resolveRepo } from '../lib/repo.ts';
import { detectScope, type Scope } from '../lib/scope.ts';

export interface PlanCommandDeps {
  client?: GraphQLClient;
  rest?: RestClient;
  hasGitRemote?: () => boolean;
  getRemoteUrl?: () => string | null;
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
  /** Override `now` for deterministic testing. */
  now?: () => Date;
  config?: AppConfig;
}

const FETCH_LIMIT = 100;
const FIELDS_FETCH_LIMIT = 50;
const DEFAULT_PERIOD: Period = 'weekly';

/**
 * Plan a daily / weekly / sprint cycle.
 *
 * - `--scope repo` (default when in a Git repo): finds-or-creates the
 *   Milestone whose title matches the period and binds open Issues whose
 *   `updatedAt` falls in the period.
 * - `--scope org|user`: finds the Iteration on the target Projects v2 board
 *   that matches the period title (or falls back to the iteration containing
 *   the current day) and updates the Iteration field on every project item
 *   updated within the period.
 *
 * `--dry-run` keeps preview output without mutating anything in either mode.
 *
 * Items already bound to a different Milestone (repo) or already on the
 * target Iteration (project) are skipped — we never silently re-route
 * someone else's planning state and we avoid no-op writes.
 */
export async function plan(argv: readonly string[], deps: PlanCommandDeps = {}): Promise<number> {
  const stdout = deps.stdout ?? process.stdout;
  const stderr = deps.stderr ?? process.stderr;
  const locale = resolveLocale(argv, process.env, deps.config);

  const period = parsePeriodFlag(argv) ?? DEFAULT_PERIOD;
  const dryRun = argv.includes('--dry-run');
  const now = deps.now ? deps.now() : new Date();

  const scope = detectScope({ argv, hasGitRemote: deps.hasGitRemote, config: deps.config });

  if (scope === 'repo') {
    return await planRepo({ argv, deps, locale, stdout, stderr, period, dryRun, now });
  }

  return await planProject({ argv, deps, locale, scope, stdout, stderr, period, dryRun, now });
}

interface PlanRepoContext {
  argv: readonly string[];
  deps: PlanCommandDeps;
  locale: 'ja' | 'en';
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  period: Period;
  dryRun: boolean;
  now: Date;
}

async function planRepo(ctx: PlanRepoContext): Promise<number> {
  const { argv, deps, locale, stdout, stderr, period, dryRun, now } = ctx;

  const repo = resolveRepo({ argv, getRemoteUrl: deps.getRemoteUrl });
  const needsToken = !deps.client || (!dryRun && !deps.rest);
  const token = needsToken ? resolveToken() : '';
  const client = deps.client ?? createClient(token);
  const rest = deps.rest ?? createRestClient(token);

  const issuesData = await client.request<ListRepoIssuesWithMilestoneResponse>(
    LIST_REPO_ISSUES_WITH_MILESTONE,
    { owner: repo.owner, name: repo.name, first: FETCH_LIMIT }
  );
  if (!issuesData.repository) {
    stderr.write(`repository not found: ${repo.owner}/${repo.name}\n`);
    return 1;
  }

  const range = rangeOf(period, now);
  const inRange = issuesData.repository.issues.nodes.filter((i: RepoIssueWithMilestoneNode) => {
    const updated = new Date(i.updatedAt).getTime();
    return updated >= range.start.getTime() && updated < range.end.getTime();
  });

  const title = suggestMilestoneTitle(period, now);
  stdout.write(`${t(locale, 'plan.proposed')}: ${title}\n`);
  stdout.write(
    `  ${range.start.toISOString().slice(0, 10)} → ${range.end.toISOString().slice(0, 10)}\n\n`
  );

  if (inRange.length === 0) {
    stdout.write(`${t(locale, 'plan.empty')}\n`);
    if (dryRun) stdout.write(`\n${t(locale, 'plan.dryRunNote')}\n`);
    return 0;
  }

  stdout.write(`${t(locale, 'plan.candidates')} (${inRange.length})\n`);
  for (const issue of inRange) {
    stdout.write(`  #${issue.number}  ${issue.title}\n`);
  }
  stdout.write('\n');

  if (dryRun) {
    stdout.write(`${t(locale, 'plan.dryRunNote')}\n`);
    return 0;
  }

  // Reuse a same-titled Milestone if present; otherwise create one.
  const milestonesData = await client.request<ListMilestonesResponse>(LIST_MILESTONES, {
    owner: repo.owner,
    name: repo.name,
    first: FETCH_LIMIT,
  });
  const existing = milestonesData.repository?.milestones.nodes.find((m) => m.title === title);

  let milestoneId: string;
  let milestoneNumber: number;
  if (existing) {
    milestoneId = existing.id;
    milestoneNumber = existing.number;
    stdout.write(`${t(locale, 'plan.reused')}: ${title} (#${existing.number})\n`);
  } else {
    const created = await createMilestone(rest, {
      owner: repo.owner,
      name: repo.name,
      title,
    });
    milestoneId = created.node_id;
    milestoneNumber = created.number;
    stdout.write(`${t(locale, 'plan.created')}: ${title} (#${created.number})\n`);
  }

  for (const issue of inRange) {
    if (issue.milestone && issue.milestone.id !== milestoneId) {
      stdout.write(
        `  ${t(locale, 'plan.skippedExisting')}: #${issue.number} → ${issue.milestone.title}\n`
      );
      continue;
    }
    if (issue.milestone && issue.milestone.id === milestoneId) {
      // Already bound to the right milestone — nothing to do.
      continue;
    }
    await client.request<UpdateIssueMilestoneResponse>(UPDATE_ISSUE_MILESTONE, {
      input: { id: issue.id, milestoneId },
    });
    stdout.write(`  ${t(locale, 'plan.linked')}: #${issue.number}\n`);
  }

  // Surface the Milestone number so the user can jump to it without scanning.
  stdout.write(`\nhttps://github.com/${repo.owner}/${repo.name}/milestone/${milestoneNumber}\n`);
  return 0;
}

interface PlanProjectContext {
  argv: readonly string[];
  deps: PlanCommandDeps;
  locale: 'ja' | 'en';
  scope: Exclude<Scope, 'repo'>;
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  period: Period;
  dryRun: boolean;
  now: Date;
}

async function planProject(ctx: PlanProjectContext): Promise<number> {
  const { argv, deps, locale, scope, stdout, stderr, period, dryRun, now } = ctx;

  let projectRef: ProjectRef;
  try {
    projectRef = resolveProjectRef({ scope, argv, config: deps.config });
  } catch (err) {
    if (err instanceof ProjectError) {
      stderr.write(`${err.message}\n`);
      return 2;
    }
    throw err;
  }

  const client = deps.client ?? createClient(resolveToken());
  const projectId = await resolveProjectNodeId({ client, scope, projectRef });
  if (!projectId) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const fieldsData = await client.request<ListProjectV2FieldsResponse>(LIST_PROJECT_V2_FIELDS, {
    projectId,
    first: FIELDS_FETCH_LIMIT,
  });
  if (!fieldsData.node) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const iterationField = findIterationField(fieldsData.node.fields.nodes);
  if (!iterationField) {
    stderr.write(`${t(locale, 'error.plan.iterationFieldMissing')}\n`);
    return 1;
  }

  const range = rangeOf(period, now);
  const targetTitle = suggestMilestoneTitle(period, now);

  const resolved = resolveTargetIteration(iterationField.iterations, targetTitle, now);
  if (!resolved) {
    stderr.write(`${t(locale, 'error.plan.noIterationsAvailable')}\n`);
    return 1;
  }

  // Header: proposed iteration + period range.
  stdout.write(`${t(locale, 'plan.proposed.project')}: ${resolved.iteration.title}\n`);
  stdout.write(
    `  ${range.start.toISOString().slice(0, 10)} → ${range.end.toISOString().slice(0, 10)}\n`
  );
  if (resolved.matched) {
    stdout.write(`  ${t(locale, 'plan.iterationMatched.project')}: ${targetTitle}\n\n`);
  } else {
    // Surface fallback to stderr so scripts can detect it; also echo to stdout
    // for the human reading the preview.
    stderr.write(
      `${t(locale, 'plan.iterationFallback.project')}: ${targetTitle} → ${resolved.iteration.title}\n`
    );
    stdout.write(`  ${t(locale, 'plan.iterationFallback.project')}\n\n`);
  }

  const itemsData = await client.request<ListProjectV2ItemsResponse>(LIST_PROJECT_V2_ITEMS, {
    projectId,
    first: FETCH_LIMIT,
  });
  const allItems = itemsData.node?.items.nodes ?? [];
  const inRange = allItems.filter((item) => {
    const updated = new Date(item.updatedAt).getTime();
    return updated >= range.start.getTime() && updated < range.end.getTime();
  });

  if (inRange.length === 0) {
    stdout.write(`${t(locale, 'plan.empty.project')}\n`);
    if (dryRun) stdout.write(`\n${t(locale, 'plan.dryRunNote.project')}\n`);
    return 0;
  }

  stdout.write(`${t(locale, 'plan.candidates.project')} (${inRange.length})\n`);
  for (const item of inRange) {
    stdout.write(formatItemLine(item));
  }
  stdout.write('\n');

  if (dryRun) {
    stdout.write(`${t(locale, 'plan.dryRunNote.project')}\n`);
    return 0;
  }

  for (const item of inRange) {
    if (isAlreadyOnIteration(item, iterationField.id, resolved.iteration.id)) {
      stdout.write(`  ${t(locale, 'plan.iterationAlreadySet.project')}: ${describeItem(item)}\n`);
      continue;
    }
    await client.request<UpdateProjectV2ItemFieldValueResponse>(
      UPDATE_PROJECT_V2_ITEM_FIELD_VALUE,
      {
        input: {
          projectId,
          itemId: item.id,
          fieldId: iterationField.id,
          value: { iterationId: resolved.iteration.id },
        },
      }
    );
    stdout.write(`  ${t(locale, 'plan.iterationUpdated.project')}: ${describeItem(item)}\n`);
  }

  return 0;
}

interface IterationFieldShape {
  id: string;
  name: string;
  iterations: ProjectV2IterationOption[];
}

function findIterationField(fields: readonly ProjectV2FieldNode[]): IterationFieldShape | null {
  for (const f of fields) {
    if (f.dataType === 'ITERATION' && f.name.toLowerCase() === 'iteration') {
      return {
        id: f.id,
        name: f.name,
        iterations: [...f.configuration.iterations],
      };
    }
  }
  // Fall back to any ITERATION field if no field is literally named "Iteration".
  for (const f of fields) {
    if (f.dataType === 'ITERATION') {
      return {
        id: f.id,
        name: f.name,
        iterations: [...f.configuration.iterations],
      };
    }
  }
  return null;
}

interface ResolvedIteration {
  iteration: ProjectV2IterationOption;
  matched: boolean;
}

/**
 * Pick the iteration to write. Order:
 *   1. exact-title match (`Daily YYYY-MM-DD` etc.)
 *   2. iteration whose [startDate, startDate + duration) window contains `now`
 *   3. nearest upcoming iteration by startDate
 *   4. last available iteration
 *
 * Returns null only when the field has no iterations at all (the caller
 * reports an actionable error rather than picking arbitrarily).
 *
 * GitHub's Projects v2 GraphQL API exposes iteration creation only through
 * `updateProjectV2IterationField`-style mutations on the *iteration field's
 * configuration*, not as a per-iteration create. Rather than implement that
 * indirect mutation, we keep the v0.1.0 behavior: find-or-fallback. The
 * caller can pre-create iterations in the Project UI to make the matching
 * deterministic.
 */
function resolveTargetIteration(
  iterations: readonly ProjectV2IterationOption[],
  targetTitle: string,
  now: Date
): ResolvedIteration | null {
  if (iterations.length === 0) return null;

  const exact = iterations.find((it) => it.title === targetTitle);
  if (exact) return { iteration: exact, matched: true };

  const nowMs = now.getTime();
  const current = iterations.find((it) => containsDay(it, nowMs));
  if (current) return { iteration: current, matched: false };

  const upcoming = iterations
    .filter((it) => new Date(it.startDate).getTime() >= nowMs)
    .sort((a, b) => new Date(a.startDate).getTime() - new Date(b.startDate).getTime());
  if (upcoming[0]) return { iteration: upcoming[0], matched: false };

  // No current and no upcoming — pick the most recent past one (last entry).
  const last = iterations[iterations.length - 1];
  if (last) return { iteration: last, matched: false };
  return null;
}

function containsDay(iteration: ProjectV2IterationOption, nowMs: number): boolean {
  const start = new Date(iteration.startDate).getTime();
  if (Number.isNaN(start)) return false;
  const end = start + iteration.duration * 24 * 60 * 60 * 1000;
  return nowMs >= start && nowMs < end;
}

function isAlreadyOnIteration(
  item: ProjectV2ItemNode,
  iterationFieldId: string,
  iterationId: string
): boolean {
  for (const v of item.fieldValues.nodes) {
    if (
      v.__typename === 'ProjectV2ItemFieldIterationValue' &&
      v.field.id === iterationFieldId &&
      v.iterationId === iterationId
    ) {
      return true;
    }
  }
  return false;
}

function formatItemLine(item: ProjectV2ItemNode): string {
  const c = item.content;
  if (!c) return `  (no content)\n`;
  if (c.__typename === 'Issue' || c.__typename === 'PullRequest') {
    const prefix = c.__typename === 'PullRequest' ? 'PR' : '';
    return `  ${prefix}#${c.number}  ${c.title}\n`;
  }
  return `  (draft)  ${c.title}\n`;
}

function describeItem(item: ProjectV2ItemNode): string {
  const c = item.content;
  if (!c) return item.id;
  if (c.__typename === 'Issue' || c.__typename === 'PullRequest') {
    const prefix = c.__typename === 'PullRequest' ? 'PR' : '';
    return `${prefix}#${c.number}`;
  }
  return item.id;
}

File: cmd/plan.go
package cmd

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/ozzy-labs/gh-tasks/internal/github/queries"
	"github.com/ozzy-labs/gh-tasks/internal/period"
	"github.com/ozzy-labs/gh-tasks/internal/project"
	"github.com/ozzy-labs/gh-tasks/internal/projectitem"
	"github.com/ozzy-labs/gh-tasks/internal/repo"
	"github.com/ozzy-labs/gh-tasks/internal/scope"
)

const (
	planFetchLimit       = 100
	planFieldsFetchLimit = 50
)

func newPlanCmd(deps Deps) *cobra.Command {
	c := &cobra.Command{
		Use:   "plan",
		Short: "Plan tasks for a period (daily / weekly / sprint)",
		RunE: func(c *cobra.Command, _ []string) error {
			return runPlan(c.Context(), c, deps)
		},
	}
	c.Flags().String("period", "weekly", "aggregation window: daily | weekly | sprint")
	c.Flags().Bool("write", false, "apply milestone / iteration changes (otherwise preview only)")
	addJSONFlags(c)
	addJSONCompletion(c, itemJSONFields)
	return c
}

func runPlan(ctx context.Context, c *cobra.Command, deps Deps) error {
	r, err := deps.Resolve(c)
	if err != nil {
		return localizedError(c, r, err)
	}
	jsonReq, jsonOn, err := resolveJSONRequest(c, r, itemJSONFields)
	if err != nil {
		return err
	}
	pflag, _ := c.Flags().GetString("period")
	write, _ := c.Flags().GetBool("write")
	p, err := period.Parse(pflag)
	if err != nil {
		return localizedError(c, r, err)
	}
	now := deps.Now()
	rng := period.Of(p, period.Options{Getenv: deps.Env, Now: now})
	sc, err := scope.Detect(scope.DetectOptions{
		Flag:         flagString(c, "scope"),
		HasGitRemote: deps.HasGitRemote,
		DefaultScope: r.Config.DefaultScope,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	if sc == scope.Repo {
		return runPlanRepo(ctx, c, deps, r, p, rng, write, now, jsonOn, jsonReq)
	}
	return runPlanProject(ctx, c, deps, r, sc, p, rng, write, now, jsonOn, jsonReq)
}

func runPlanRepo(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, p period.Period, rng period.Range, write bool, now time.Time, jsonOn bool, jsonReq jsonRequest) error {
	id, err := repo.Resolve(repo.ResolveOptions{Flag: flagString(c, "repo"), GetRemoteURL: deps.GetRemoteURL})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	gqlClient := clients.AsGenqlientClient()
	issues, err := queries.PaginateRepoIssuesWithMilestone(ctx, gqlClient, id.Owner, id.Name, planFetchLimit)
	if errors.Is(err, queries.ErrRepoNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.repo.notFound", "owner", id.Owner, "name", id.Name))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list repo issues with milestone", err)
	}
	inRange := []*queries.RepoIssueWithMilestone{}
	for _, n := range issues {
		if n == nil {
			continue
		}
		if t, err := time.Parse(time.RFC3339, n.UpdatedAt); err == nil &&
			(t.Equal(rng.Start) || t.After(rng.Start)) && t.Before(rng.End) {
			inRange = append(inRange, n)
		}
	}
	if jsonOn && !write {
		// preview-only path: emit the candidate list as a flat JSON
		// array. write-mode JSON falls through to the bind loop below
		// and emits the bound items at the end.
		rows := make([]map[string]any, 0, len(inRange))
		for _, n := range inRange {
			rows = append(rows, map[string]any{
				"id": n.Id, "number": n.Number, "state": "OPEN",
				"title": n.Title, "type": "ISSUE", "updatedAt": n.UpdatedAt, "url": n.Url,
			})
		}
		return renderJSONItems(c, r, rows, jsonReq, itemJSONFields)
	}
	title := period.SuggestMilestoneTitle(p, period.Options{Getenv: deps.Env, Now: now})
	out := c.OutOrStdout()
	fmt.Fprintf(out, "%s: %s\n", r.T("plan.proposed"), title)
	fmt.Fprintf(out, "  %s → %s\n\n", rng.Start.Format("2006-01-02"), rng.End.Format("2006-01-02"))
	if len(inRange) == 0 {
		fmt.Fprintln(out, r.T("plan.empty"))
		if !write {
			fmt.Fprintf(out, "\n%s\n", r.T("plan.previewNote"))
		}
		return nil
	}
	fmt.Fprintf(out, "%s (%d)\n", r.T("plan.candidates"), len(inRange))
	for _, n := range inRange {
		fmt.Fprintf(out, "  #%d  %s\n", n.Number, n.Title)
	}
	fmt.Fprintln(out)
	if !write {
		fmt.Fprintln(out, r.T("plan.previewNote"))
		return nil
	}

	milestones, err := queries.PaginateMilestones(ctx, gqlClient, id.Owner, id.Name, planFetchLimit)
	if errors.Is(err, queries.ErrRepoNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.repo.notFound", "owner", id.Owner, "name", id.Name))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list milestones", err)
	}
	var milestoneID string
	var milestoneNumber int
	for _, m := range milestones {
		if m == nil {
			continue
		}
		if m.Title == title {
			milestoneID = m.Id
			milestoneNumber = m.Number
			fmt.Fprintf(out, "%s: %s (#%d)\n", r.T("plan.reused"), title, m.Number)
			break
		}
	}
	if milestoneID == "" {
		var created queries.CreateMilestoneResult
		path := fmt.Sprintf("repos/%s/%s/milestones", id.Owner, id.Name)
		body := map[string]any{"title": title}
		if err := clients.REST.Do(ctx, "POST", path, body, &created); err != nil {
			return wrapTransport(c.ErrOrStderr(), r.Locale, "create milestone", err)
		}
		milestoneID = created.NodeID
		milestoneNumber = created.Number
		fmt.Fprintf(out, "%s: %s (#%d)\n", r.T("plan.created"), title, created.Number)
	}

	bound := []*queries.RepoIssueWithMilestone{}
	for _, n := range inRange {
		if n.Milestone != nil && n.Milestone.Id != milestoneID {
			fmt.Fprintf(out, "  %s: #%d → %s\n", r.T("plan.skippedExisting"), n.Number, n.Milestone.Title)
			continue
		}
		if n.Milestone != nil && n.Milestone.Id == milestoneID {
			bound = append(bound, n)
			continue
		}
		updateInput := queries.NewUpdateIssueInput(n.Id)
		updateInput.MilestoneId = &milestoneID
		if _, err := queries.UpdateIssueMilestone(ctx, gqlClient, updateInput); err != nil {
			return wrapTransport(c.ErrOrStderr(), r.Locale, fmt.Sprintf("update issue milestone (issue #%d)", n.Number), err)
		}
		bound = append(bound, n)
		fmt.Fprintf(out, "  %s: #%d\n", r.T("plan.linked"), n.Number)
	}
	fmt.Fprintf(out, "\nhttps://github.com/%s/%s/milestone/%d\n", id.Owner, id.Name, milestoneNumber)
	if jsonOn {
		rows := make([]map[string]any, 0, len(bound))
		for _, n := range bound {
			rows = append(rows, map[string]any{
				"id": n.Id, "number": n.Number, "state": "OPEN",
				"title": n.Title, "type": "ISSUE", "updatedAt": n.UpdatedAt, "url": n.Url,
			})
		}
		return renderJSONItems(c, r, rows, jsonReq, itemJSONFields)
	}
	return nil
}

func runPlanProject(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, sc scope.Scope, p period.Period, rng period.Range, write bool, now time.Time, jsonOn bool, jsonReq jsonRequest) error {
	pref, err := project.Resolve(project.ResolveOptions{
		Scope:       sc,
		Flag:        flagString(c, "project"),
		OrgProject:  r.Config.OrgProject,
		UserProject: r.Config.UserProject,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	pid, err := projectitem.ResolveProjectNodeID(ctx, clients.GraphQL, sc, pref)
	if err != nil {
		return localizedError(c, r, err)
	}
	if pid == "" {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	gqlClient := clients.AsGenqlientClient()
	fieldNodes, err := queries.PaginateProjectV2Fields(ctx, gqlClient, pid, planFieldsFetchLimit)
	if errors.Is(err, queries.ErrProjectNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list project fields", err)
	}
	fields := projectitem.FieldsOf(fieldNodes)
	itField := findIterationField(fields)
	if itField == nil || itField.Configuration == nil {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.plan.iterationFieldMissing"))
		return ErrSilentRuntime
	}
	target := period.SuggestMilestoneTitle(p, period.Options{Getenv: deps.Env, Now: now})
	resolved := resolveTargetIteration(itField.Configuration.Iterations, target, now)
	if resolved == nil {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.plan.noIterationsAvailable"))
		return ErrSilentRuntime
	}
	out := c.OutOrStdout()
	fmt.Fprintf(out, "%s: %s\n", r.T("plan.proposed.project"), resolved.iteration.Title)
	fmt.Fprintf(out, "  %s → %s\n", rng.Start.Format("2006-01-02"), rng.End.Format("2006-01-02"))
	if resolved.matched {
		fmt.Fprintf(out, "  %s: %s\n\n", r.T("plan.iterationMatched.project"), target)
	} else {
		fmt.Fprintf(c.ErrOrStderr(), "%s: %s → %s\n", r.T("plan.iterationFallback.project"), target, resolved.iteration.Title)
		fmt.Fprintf(out, "  %s\n\n", r.T("plan.iterationFallback.project"))
	}

	allItems, err := queries.PaginateProjectV2Items(ctx, gqlClient, pid, planFetchLimit)
	if errors.Is(err, queries.ErrProjectNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list project items", err)
	}
	inRange := []*queries.ProjectV2ItemNode{}
	for _, item := range allItems {
		if item == nil {
			continue
		}
		if t, err := time.Parse(time.RFC3339, item.UpdatedAt); err == nil &&
			(t.Equal(rng.Start) || t.After(rng.Start)) && t.Before(rng.End) {
			inRange = append(inRange, item)
		}
	}
	if jsonOn && !write {
		// preview-only path (see runPlanRepo for rationale).
		return renderJSONItems(c, r, projectItemRowsToJSON(inRange), jsonReq, itemJSONFields)
	}
	if len(inRange) == 0 {
		fmt.Fprintln(out, r.T("plan.empty.project"))
		if !write {
			fmt.Fprintf(out, "\n%s\n", r.T("plan.previewNote.project"))
		}
		return nil
	}
	fmt.Fprintf(out, "%s (%d)\n", r.T("plan.candidates.project"), len(inRange))
	for _, item := range inRange {
		fmt.Fprint(out, formatItemLineForPlan(item))
	}
	fmt.Fprintln(out)
	if !write {
		fmt.Fprintln(out, r.T("plan.previewNote.project"))
		return nil
	}
	bound := []*queries.ProjectV2ItemNode{}
	for _, item := range inRange {
		if isAlreadyOnIteration(item, itField.ID, resolved.iteration.ID) {
			fmt.Fprintf(out, "  %s: %s\n", r.T("plan.iterationAlreadySet.project"), describeItem(item))
			bound = append(bound, item)
			continue
		}
		if _, err := queries.UpdateProjectV2ItemFieldValue(ctx, gqlClient, &queries.UpdateProjectV2ItemFieldValueInput{
			ProjectId: pid,
			ItemId:    item.Id,
			FieldId:   itField.ID,
			Value:     &queries.ProjectV2FieldValue{IterationId: &resolved.iteration.ID},
		}); err != nil {
			return wrapTransport(c.ErrOrStderr(), r.Locale, fmt.Sprintf("update item field value (%s)", describeItem(item)), err)
		}
		fmt.Fprintf(out, "  %s: %s\n", r.T("plan.iterationUpdated.project"), describeItem(item))
		bound = append(bound, item)
	}
	if jsonOn {
		return renderJSONItems(c, r, projectItemRowsToJSON(bound), jsonReq, itemJSONFields)
	}
	return nil
}

func findIterationField(fields []projectitem.FieldDescriptor) *projectitem.FieldDescriptor {
	for i := range fields {
		f := &fields[i]
		if f.DataType == "ITERATION" && strings.EqualFold(f.Name, "iteration") {
			return f
		}
	}
	for i := range fields {
		f := &fields[i]
		if f.DataType == "ITERATION" {
			return f
		}
	}
	return nil
}

type resolvedIteration struct {
	iteration projectitem.IterationOption
	matched   bool
}

func resolveTargetIteration(iterations []projectitem.IterationOption, target string, now time.Time) *resolvedIteration {
	if len(iterations) == 0 {
		return nil
	}
	for _, it := range iterations {
		if it.Title == target {
			return &resolvedIteration{iteration: it, matched: true}
		}
	}
	for _, it := range iterations {
		if iterationContains(it, now) {
			return &resolvedIteration{iteration: it, matched: false}
		}
	}
	type parsedIteration struct {
		iteration projectitem.IterationOption
		start     time.Time
	}
	upcoming := []parsedIteration{}
	for _, it := range iterations {
		start, err := parseIterationStart(it.StartDate)
		if err != nil {
			continue
		}
		if start.Equal(now) || start.After(now) {
			upcoming = append(upcoming, parsedIteration{iteration: it, start: start})
		}
	}
	// Use SliceStable to preserve catalog order for iterations sharing a start
	// date; the parsed time is captured at filter time, so the comparator never
	// re-parses (and therefore never silently treats a parse error as epoch).
	sort.SliceStable(upcoming, func(i, j int) bool {
		return upcoming[i].start.Before(upcoming[j].start)
	})
	if len(upcoming) > 0 {
		return &resolvedIteration{iteration: upcoming[0].iteration, matched: false}
	}
	return &resolvedIteration{iteration: iterations[len(iterations)-1], matched: false}
}

func iterationContains(it projectitem.IterationOption, now time.Time) bool {
	start, err := parseIterationStart(it.StartDate)
	if err != nil {
		return false
	}
	end := start.AddDate(0, 0, it.Duration)
	return (now.Equal(start) || now.After(start)) && now.Before(end)
}

func parseIterationStart(s string) (time.Time, error) {
	if t, err := time.Parse("2006-01-02", s); err == nil {
		return t, nil
	}
	return time.Parse(time.RFC3339, s)
}

func isAlreadyOnIteration(item *queries.ProjectV2ItemNode, fieldID, iterationID string) bool {
	for _, v := range projectitem.FieldValuesOf(item) {
		if v.Typename == "ProjectV2ItemFieldIterationValue" && v.Field.ID == fieldID && v.IterationID == iterationID {
			return true
		}
	}
	return false
}

func formatItemLineForPlan(item *queries.ProjectV2ItemNode) string {
	c := projectitem.ContentOf(item)
	switch c.Typename {
	case "Issue":
		return fmt.Sprintf("  #%d  %s\n", c.Number, c.Title)
	case "PullRequest":
		return fmt.Sprintf("  PR#%d  %s\n", c.Number, c.Title)
	case "DraftIssue":
		return fmt.Sprintf("  (draft)  %s\n", c.Title)
	default:
		return "  (no content)\n"
	}
}

func describeItem(item *queries.ProjectV2ItemNode) string {
	c := projectitem.ContentOf(item)
	switch c.Typename {
	case "Issue":
		return fmt.Sprintf("#%d", c.Number)
	case "PullRequest":
		return fmt.Sprintf("PR#%d", c.Number)
	default:
		if item == nil {
			return ""
		}
		return item.Id
	}
}
